File: qbixora.py
#!/usr/bin/env python

'''
//  -------------------------------------------------------------
//  author        Giga
//  project       qeeqbox/ixora
//  email         gigaqeeq@gmail.com
//  description   app.py (CLI)
//  -------------------------------------------------------------
//  contributors list qeeqbox/ixora/graphs/contributors
//  -------------------------------------------------------------
'''

from jinja2 import Template, Environment, FileSystemLoader
from webbrowser import open as wopen
from os import path

class QBIxora:
    def __init__(self,project_name,verbose=False):
        self.project_name = project_name
        self.verbose = verbose;
        self.graph = {'nodes':[],'links':[],'search_input':[],'search_index':[]}
        self.temp_nodes = []
        self.temp_edges = []
        self.graph_py = path.join(path.dirname(__file__), "data", "graph_py.html")
        self.base_html_file = open(self.graph_py,encoding='utf-8').read()

    def add_node(self,name,search=None,_set=None):
        if name not in self.temp_nodes:
            self.temp_nodes.append(name)
            temp_node = {'name': name,'_id':len(self.graph['nodes'])}
            if _set != None:
                temp_node.update(_set)
            self.graph['nodes'].append(temp_node)
        if search not in self.graph['search_input']:
            self.graph['search_input'].append(search)
            self.graph['search_index'].append(self.temp_nodes.index(name))

    def add_edge(self,src,dst,_set=None):
        src_dst = '{}:{}'.format(src,dst)
        dst_src = '{}:{}'.format(dst,src)
        if src_dst not in self.temp_edges and dst_src not in self.temp_edges:
            try:
                _src = self.temp_nodes.index(src)
                _dst = self.temp_nodes.index(dst)
                if _src != None and _dst != None:
                    self.temp_edges.append(src_dst)
                    temp_edge = {'source': _src, 'target': _dst}
                    if _set != None:
                        temp_edge.update(_set)
                    self.graph['links'].append(temp_edge)
            except:
                pass

    def create_graph(self,div,window_title='',search_title='',search_msg='', copyright_link='',copyright_msg='',tools=[], collide=10,distance=100,data={},method='',save_to=None,open_file=False):
        current_subject = False;
        right_click_node = None
        override_defualt_node_color = None
        override_defualt_link_color = None
        defualt_node_color = '#D4D4D4'
        defualt_edge_color = '#b2b2b2'
        search_on_off_rendered= True if 'search' in tools else False
        tooltop_on_off_rendered = True if 'tooltip' in tools else False
        window_on_off_rendered = True if 'window' in tools else False
        search_title_rendered = search_title
        search_msg_rendered = search_msg
        tools_rendered = tools
        collide_rendered = collide
        distance_rendered = distance
        copyright_link_rendered = copyright_link
        copyright_msg_rendered = copyright_msg

        if len(data) > 0:
            self.graph = data

        ixora_object = {'graph_div':div,'graph':self.graph,'current_subject':current_subject,'right_click_node':right_click_node,'override_defualt_node_color':override_defualt_node_color,'override_defualt_link_color':override_defualt_link_color,'defualt_node_color':defualt_node_color,'defualt_edge_color':defualt_edge_color,'search_on_off_rendered':search_on_off_rendered,'tooltop_on_off_rendered':tooltop_on_off_rendered,'search_title_rendered':search_title_rendered,'search_msg_rendered':search_msg_rendered,'tools_rendered':tools_rendered,'window_on_off_rendered':window_on_off_rendered,'collide_rendered':collide_rendered,'distance_rendered':distance_rendered,'copyright_link_rendered':copyright_link_rendered,'copyright_msg_rendered':copyright_msg_rendered}

        if method == 'object':
            return ixora_object

        if method == 'file_with_json':
            rendered = Template(self.base_html_file).render(project_name=self.project_name,graph=ixora_object)
            with open(save_to, 'w', encoding='utf-8') as f:
                f.write(rendered)
            if path.exists(save_to):
                if open_file:
                    wopen(save_to)
                return True
        return False

File: qbixora.js
//  -------------------------------------------------------------
//  author        Giga
//  project       qeeqbox/ixora
//  email         gigaqeeq@gmail.com
//  description   app.py (CLI)
//  -------------------------------------------------------------
//  contributors list qeeqbox/ixora/graphs/contributors
//  -------------------------------------------------------------

var ejs  = require('ejs');
var fs   = require('fs');
var open = require('open');
var path = require('path')

class QBIxora {
  constructor(project_name,verbose=false) {
    this.project_name = project_name
    this.verbose = verbose;
    this.graph = {
      'nodes': [],
      'links': [],
      'search_input': [],
      'search_index': []
    }
    this.temp_nodes = []
    this.temp_edges = []
    this.base_html_file = fs.readFileSync(path.join(__dirname,'graph_js.html'), 'utf-8');
    this.base_html = ejs.render ( this.base_html_file , {project_name:this.project_name});
  }

  add_node(name, search = undefined, _set = undefined) {
    try{
      if (this.temp_nodes.length < 3000){
        if (typeof name !== 'undefined' && name !== null && !this.temp_nodes.includes(name)) {
          this.temp_nodes.push(name)
          var temp_node = {
            "name": name,
            '_id': this.graph['nodes'].length
          }
          if (typeof _set !== 'undefined' && _set !== null) {
            temp_node = Object.assign({}, _set, temp_node)
          }
          this.graph['nodes'].push(temp_node)
        }
        if (typeof search !== 'undefined' && search !== null && !this.graph['search_input'].includes(search)) {
          this.graph['search_input'].push(search)
          this.graph['search_index'].push(this.temp_nodes.indexOf(name))
        }
      }
    }
    catch (err){
      this.verbose && console.log(err);
    }
  }

  add_edge(src, dst, _set = undefined) {
    var src_dst = src + ":" + dst
    var dst_src = dst + ":" + src
    if (typeof src_dst !== 'undefined' && src_dst !== null && !this.temp_edges.includes(src_dst) && typeof dst_src !== 'undefined' && dst_src !== null && !this.temp_edges.includes(dst_src))
      try {
        var _src = this.temp_nodes.indexOf(src)
        var _dst = this.temp_nodes.indexOf(dst)
        if (_src != -1 && _dst != -1) {
          this.temp_edges.push(src_dst)
          var temp_edge = {
            "source": _src,
            "target": _dst
          }
          if (typeof _set !== 'undefined' && _set !== null) {
            temp_edge = Object.assign({}, _set, temp_edge)
          }

          this.graph['links'].push(temp_edge)
        }
      }
      catch (err){
        this.verbose && console.log(err);
      }
  }
  create_graph(div,window_title="",search_title="",search_msg="", copyright_link="",copyright_msg="",tools=[], collide=10,distance=100,data={},method="",save_to=undefined,open_file=false)
  {

    var current_subject = false;
    var right_click_node = undefined
    var override_defualt_node_color = undefined
    var override_defualt_link_color = undefined
    var defualt_node_color = '#D4D4D4'
    var defualt_edge_color = '#b2b2b2'
    var search_on_off_rendered= (tools.includes('search')  ? true : false);
    var tooltop_on_off_rendered = (tools.includes('tooltip')  ? true : false);
    var window_on_off_rendered = (tools.includes('window')  ? true : false);
    var search_title_rendered = search_title
    var search_msg_rendered = search_msg
    var tools_rendered = tools
    var collide_rendered = collide
    var distance_rendered = distance
    var copyright_link_rendered = copyright_link
    var copyright_msg_rendered = copyright_msg

    try{
      if (Object.keys(data).length !== 0){
        this.graph = data
      }

      var ixora_object = {graph_div:div,graph:this.graph,current_subject:current_subject,right_click_node:right_click_node,override_defualt_node_color:override_defualt_node_color,override_defualt_link_color:override_defualt_link_color,defualt_node_color:defualt_node_color,defualt_edge_color:defualt_edge_color,search_on_off_rendered:search_on_off_rendered,tooltop_on_off_rendered:tooltop_on_off_rendered,search_title_rendered:search_title_rendered,search_msg_rendered:search_msg_rendered,tools_rendered:tools_rendered,window_on_off_rendered:window_on_off_rendered,collide_rendered:collide_rendered,distance_rendered:distance_rendered,copyright_link_rendered:copyright_link_rendered,copyright_msg_rendered:copyright_msg_rendered}

      if (method == 'object'){
        return ixora_object
      }
      else{
        var template = fs.readFileSync('template_js.html', 'utf-8');
        var rendered = ejs.render ( template , {ixora_object:ixora_object});
        if (typeof save_to !== 'undefined' && save_to !== null && rendered != ""){
            fs.writeFileSync(save_to, rendered, 'utf8');
            if (open_file){
                open(save_to);
            }
        }
        else{
          return rendered
        }
      }
    }
    catch (err){
      this.verbose && console.log(err);
    }

    return undefined
  }

  save_base_html(location){
    try{
      fs.writeFileSync(location, this.base_html);
    }
    catch (err){
      this.verbose && console.log(err);
    }
  }
}

module.exports = {
    QBIxora
};
